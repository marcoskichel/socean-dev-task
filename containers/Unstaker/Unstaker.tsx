import { styled } from '@mui/material';
import { useMemo, useState } from 'react';
import FatButton from '../../components/FatButton/FatButton';
import StakeInput from '../../components/StakeInput/StakeInput';
import UnstakeOptionList from '../../components/UnstakeOptionList/UnstakeOptionList';
import { Currency } from '../../types/Currency';
import { UnstakeOption } from '../../types/UnstakeOption';

const Root = styled('div')(() => ({
  display: 'flex',
  flexDirection: 'column',
  padding: '1.5rem',
  gap: '2rem',
}));

const inferOutcomeOptionFee = (
  opt: Partial<UnstakeOption>,
  stakedValue: number,
) => {
  const actualFee = opt.fee! / 100;
  const formatedFee = 1 - actualFee;
  return {
    ...opt,
    value: stakedValue * formatedFee,
  } as UnstakeOption;
};

const options: Array<Partial<UnstakeOption>> = [
  { id: 'instant', label: 'Instant Unstake', fee: 0.3 },
  { id: 'two-days', label: 'Unstake in ~2 days', fee: 0 },
];

const Unstaker = () => {
  const [amount, setAmount] = useState('0');

  const parsedOptions = useMemo(() => {
    const stakedValue = Number(amount) || 0;
    return options.map((opt) => inferOutcomeOptionFee(opt, stakedValue));
  }, [amount]);

  const [selectedOptionId, setSelectedOptionId] = useState(
    options[options.length - 1].id,
  );

  const handleUnstakeOptionSelect = (optionId: string) => {
    setSelectedOptionId(optionId);
  };

  const handleAmountChange = (value: string) => {
    setAmount(value);
  };

  return (
    <Root>
      <StakeInput
        currency={Currency.scnSOL}
        value={amount}
        onChange={handleAmountChange}
      />
      <UnstakeOptionList
        selectedId={selectedOptionId!}
        onChange={handleUnstakeOptionSelect}
        options={parsedOptions}
      />
      <FatButton
        color="secondary"
        variant="contained"
        size="large"
        disableElevation
      >
        Unstake scnSOL
      </FatButton>
    </Root>
  );
};

export default Unstaker;
