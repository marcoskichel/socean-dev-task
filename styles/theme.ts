import { createTheme, PaletteOptions, Theme } from '@mui/material';

export type CustomPallete = PaletteOptions & { bg: string };

const theme: Theme = createTheme({
  typography: {
    fontFamily: '"Manrope", "Helvetica", "Arial", sans-serif',
    h1: {
      fontWeight: 700,
      fontSize: '2.5rem',
      marginTop: '6rem',
      marginBottom: '1rem',
    },
    h6: {
      fontWeight: 700,
      fontSize: '1rem',
      lineHeight: '1.5rem',
      letter: '-1%',
    },
    body1: {
      fontWeight: 400,
      fontSize: '0.875rem',
      lineHeight: '1rem',
    },
    button: {
      textTransform: 'none',
    },
  },
  components: {
    MuiCard: {
      defaultProps: {
        sx: {
          borderRadius: '1.5rem',
        },
      },
    },
  },
  palette: {
    mode: 'dark',
    background: {
      paper: '#232323',
    },
    primary: {
      main: '#FFFFFF',
    },
    secondary: {
      main: '#F24F83',
    },
    text: {
      primary: '#FFFFFF',
      secondary: '#A7A7A7',
    },
    action: {
      disabled: '#5C5C5C',
    },
  },
});

export default theme;
