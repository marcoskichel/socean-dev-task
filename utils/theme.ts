import { createTheme, Theme } from "@mui/material";
import { red } from "@mui/material/colors";

const theme: Theme = createTheme({
  palette: {
    primary: {
      main: '#FFFFFF',
    },
    secondary: {
      main: '#A7A7A7',
    },
    action: {
      disabled: '#5C5C5C'
    }
  },
});

export default theme;