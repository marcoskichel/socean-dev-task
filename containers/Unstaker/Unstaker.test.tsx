import { ThemeProvider } from '@mui/material';
import { render, screen } from '@testing-library/react';
import userEvent from '@testing-library/user-event';
import theme from '../../styles/theme';
import Unstaker from './Unstaker';

describe('Unstaker', () => {
  it('renders', () => {
    const { asFragment } = render(
      <ThemeProvider theme={theme}>
        <Unstaker />
      </ThemeProvider>,
    );
    expect(asFragment()).toMatchSnapshot();
  });

  it('update the unstake options', async () => {
    render(
      <ThemeProvider theme={theme}>
        <Unstaker />
      </ThemeProvider>,
    );
    const input = screen.getByDisplayValue('0');
    userEvent.type(input, '1');

    const opt1 = await screen.findByText('0.997 SOL');
    const opt2 = await screen.findByText('1 SOL');
    expect(opt1).toBeInTheDocument();
    expect(opt2).toBeInTheDocument();
  });
});
