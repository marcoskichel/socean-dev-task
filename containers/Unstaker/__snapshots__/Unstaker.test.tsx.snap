// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Unstaker renders 1`] = `
<DocumentFragment>
  <div
    class="css-64g5hu"
  >
    <div
      class="css-1jke4yk"
    >
      <label
        class="MuiInputLabel-root MuiInputLabel-animated MuiFormLabel-root MuiFormLabel-colorPrimary css-1j0ijkw-MuiFormLabel-root-MuiInputLabel-root"
      >
        Amount
      </label>
      <div
        class="MuiFormControl-root MuiTextField-root css-1ybss1w-MuiFormControl-root-MuiTextField-root"
      >
        <div
          class="MuiOutlinedInput-root MuiInputBase-root MuiInputBase-colorPrimary MuiInputBase-formControl MuiInputBase-adornedStart MuiInputBase-adornedEnd css-oobokm-MuiInputBase-root-MuiOutlinedInput-root"
        >
          <div
            class="css-1wowqjn"
          >
            <span
              style="box-sizing: border-box; display: inline-block; overflow: hidden; width: 32px; height: 32px; background: none; opacity: 1; border: 0px; margin: 0px; padding: 0px; position: relative;"
            >
              <img
                alt="Currency"
                data-nimg="fixed"
                decoding="async"
                src="data:image/gif;base64,R0lGODlhAQABAIAAAAAAAP///yH5BAEAAAAALAAAAAABAAEAAAIBRAA7"
                style="position: absolute; top: 0px; left: 0px; bottom: 0px; right: 0px; box-sizing: border-box; padding: 0px; margin: auto; display: block; width: 0px; height: 0px; min-width: 100%; max-width: 100%; min-height: 100%; max-height: 100%;"
              />
              <noscript />
            </span>
          </div>
          <input
            aria-invalid="false"
            class="MuiOutlinedInput-input MuiInputBase-input MuiInputBase-inputAdornedStart MuiInputBase-inputAdornedEnd css-vxvc7x-MuiInputBase-input-MuiOutlinedInput-input"
            id="mui-1"
            type="number"
            value="0"
          />
          <div
            class="css-1dqf7fm"
          >
            <span
              class="MuiTypography-root MuiTypography-caption css-1ajvh0m-MuiTypography-root"
            >
              Balance: -
            </span>
            <button
              class="MuiButton-root MuiButton-contained MuiButton-containedSecondary MuiButton-sizeSmall MuiButton-containedSizeSmall MuiButton-disableElevation MuiButtonBase-root  css-tf6vm2-MuiButtonBase-root-MuiButton-root"
              tabindex="0"
              type="button"
            >
              MAX
              <span
                class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
              />
            </button>
          </div>
          <fieldset
            aria-hidden="true"
            class="MuiOutlinedInput-notchedOutline css-9425fu-MuiOutlinedInput-notchedOutline"
          >
            <legend
              class="css-nnbavb"
            >
              <span
                class="notranslate"
              >
                â€‹
              </span>
            </legend>
          </fieldset>
        </div>
      </div>
      <div
        class="css-vvfmul"
      >
        <svg
          aria-hidden="true"
          class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-i4bv87-MuiSvgIcon-root"
          data-testid="ArrowDownwardIcon"
          focusable="false"
          viewBox="0 0 24 24"
        >
          <path
            d="m20 12-1.41-1.41L13 16.17V4h-2v12.17l-5.58-5.59L4 12l8 8 8-8z"
          />
        </svg>
      </div>
    </div>
    <ul
      class="MuiList-root MuiList-padding css-h4y409-MuiList-root"
    >
      <div
        class="MuiListItemButton-root MuiListItemButton-gutters MuiButtonBase-root css-1n62zx8-MuiButtonBase-root-MuiListItemButton-root"
        role="button"
        tabindex="0"
      >
        <div
          class="css-1kxonj9"
        >
          <div
            class="css-19s3t1h"
          >
            <span
              class="MuiTypography-root MuiTypography-caption css-1kpht1s-MuiTypography-root"
            >
              Instant Unstake
            </span>
            <div
              class="css-70qvj9"
            >
              <div
                class="css-cleg7o"
              >
                <span
                  style="box-sizing: border-box; display: inline-block; overflow: hidden; width: 26px; height: 26px; background: none; opacity: 1; border: 0px; margin: 0px; padding: 0px; position: relative;"
                >
                  <img
                    alt="Currency"
                    data-nimg="fixed"
                    decoding="async"
                    src="data:image/gif;base64,R0lGODlhAQABAIAAAAAAAP///yH5BAEAAAAALAAAAAABAAEAAAIBRAA7"
                    style="position: absolute; top: 0px; left: 0px; bottom: 0px; right: 0px; box-sizing: border-box; padding: 0px; margin: auto; display: block; width: 0px; height: 0px; min-width: 100%; max-width: 100%; min-height: 100%; max-height: 100%;"
                  />
                  <noscript />
                </span>
              </div>
              <h5
                class="MuiTypography-root MuiTypography-h5 css-1yum79a-MuiTypography-root"
              >
                0 SOL
              </h5>
              <span
                class="MuiTypography-root MuiTypography-caption css-1kpht1s-MuiTypography-root"
              >
                0.3% fee
              </span>
            </div>
          </div>
          <span
            class="MuiCheckbox-root MuiCheckbox-colorSecondary MuiButtonBase-root MuiCheckbox-root MuiCheckbox-colorSecondary PrivateSwitchBase-root css-1tsckva-MuiButtonBase-root-MuiCheckbox-root"
          >
            <input
              class="PrivateSwitchBase-input css-1m9pwf3"
              data-indeterminate="false"
              type="checkbox"
            />
            <svg
              aria-hidden="true"
              class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-i4bv87-MuiSvgIcon-root"
              data-testid="CheckBoxOutlineBlankIcon"
              focusable="false"
              viewBox="0 0 24 24"
            >
              <path
                d="M19 5v14H5V5h14m0-2H5c-1.1 0-2 .9-2 2v14c0 1.1.9 2 2 2h14c1.1 0 2-.9 2-2V5c0-1.1-.9-2-2-2z"
              />
            </svg>
            <span
              class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
            />
          </span>
        </div>
        <span
          class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
        />
      </div>
      <div
        class="MuiListItemButton-root MuiListItemButton-gutters Mui-selected MuiButtonBase-root css-1n62zx8-MuiButtonBase-root-MuiListItemButton-root"
        role="button"
        tabindex="0"
      >
        <div
          class="css-1kxonj9"
        >
          <div
            class="css-19s3t1h"
          >
            <span
              class="MuiTypography-root MuiTypography-caption css-1kpht1s-MuiTypography-root"
            >
              Unstake in ~2 days
            </span>
            <div
              class="css-70qvj9"
            >
              <div
                class="css-cleg7o"
              >
                <span
                  style="box-sizing: border-box; display: inline-block; overflow: hidden; width: 26px; height: 26px; background: none; opacity: 1; border: 0px; margin: 0px; padding: 0px; position: relative;"
                >
                  <img
                    alt="Currency"
                    data-nimg="fixed"
                    decoding="async"
                    src="data:image/gif;base64,R0lGODlhAQABAIAAAAAAAP///yH5BAEAAAAALAAAAAABAAEAAAIBRAA7"
                    style="position: absolute; top: 0px; left: 0px; bottom: 0px; right: 0px; box-sizing: border-box; padding: 0px; margin: auto; display: block; width: 0px; height: 0px; min-width: 100%; max-width: 100%; min-height: 100%; max-height: 100%;"
                  />
                  <noscript />
                </span>
              </div>
              <h5
                class="MuiTypography-root MuiTypography-h5 css-1yum79a-MuiTypography-root"
              >
                0 SOL
              </h5>
              <span
                class="MuiTypography-root MuiTypography-caption css-1kpht1s-MuiTypography-root"
              >
                0% fee
              </span>
            </div>
          </div>
          <span
            class="MuiCheckbox-root MuiCheckbox-colorSecondary MuiButtonBase-root MuiCheckbox-root MuiCheckbox-colorSecondary PrivateSwitchBase-root Mui-checked css-1tsckva-MuiButtonBase-root-MuiCheckbox-root"
          >
            <input
              checked=""
              class="PrivateSwitchBase-input css-1m9pwf3"
              data-indeterminate="false"
              type="checkbox"
            />
            <svg
              aria-hidden="true"
              class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-i4bv87-MuiSvgIcon-root"
              data-testid="CheckBoxIcon"
              focusable="false"
              viewBox="0 0 24 24"
            >
              <path
                d="M19 3H5c-1.11 0-2 .9-2 2v14c0 1.1.89 2 2 2h14c1.11 0 2-.9 2-2V5c0-1.1-.89-2-2-2zm-9 14l-5-5 1.41-1.41L10 14.17l7.59-7.59L19 8l-9 9z"
              />
            </svg>
            <span
              class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
            />
          </span>
        </div>
        <span
          class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
        />
      </div>
    </ul>
    <button
      class="MuiButton-root MuiButton-contained MuiButton-containedSecondary MuiButton-sizeLarge MuiButton-containedSizeLarge MuiButton-disableElevation MuiButtonBase-root css-lmekah-MuiButtonBase-root-MuiButton-root"
      tabindex="0"
      type="button"
    >
      Unstake scnSOL
      <span
        class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
      />
    </button>
  </div>
</DocumentFragment>
`;
